libs/domain/feature-loaders/src/snapshot/SnapshotProposalLoader.ts

```diff
@@ -0,0 +1,187 @@
+        name:t.string,
+    })),()=>"strategies failed"),
+    title: t.string,
+    body: optional(t.string),
+    choices: t.array(t.string),
+    start: t.number,
+    end: t.number,
+    snapshot: t.string,
+    state: t.string,
+    link: optional(t.string),
+    scores: optional(t.array(t.number)),
+    votes: optional(t.number)
+});
+
+export const ProposalsCodec = t.strict({
+    proposals: t.array(ProposalCodec),
+});
+
+export type Proposals = t.TypeOf<typeof ProposalsCodec>;
+
+const INFO = {
+    namespace: "snapshot",
+    name: "Proposal",
+    version: "V1",
+};
+@Nested()
+class Space {
+    @OptionalProperty()
+    name: string
+    @NonEmptyProperty()
+    id: string
+}
+
+@Nested()
+class Strategy {
+    @NonEmptyProperty()
+    name:string
+}
+
+@Data({
+    info: INFO,
+})
+export class Proposal {
+    @NonEmptyProperty()
+    id: string
+    @NonEmptyProperty()
+    author: string
+    @NonEmptyProperty()
+    created: number
+    @OptionalObjectRef(Space)
+    space: Space | undefined
+    @OptionalProperty()
+    type:string | undefined
+    @RequiredArrayRef(Strategy)
+    strategies: Strategy[]
+    @NonEmptyProperty()
+    title:string
+    @OptionalProperty()
+    body: string | undefined
+    @RequiredStringArrayOf()
+    choices: string[]
+    @NonEmptyProperty()
+    start: number
+    @NonEmptyProperty()
+    end: number
+    @NonEmptyProperty()
+    snapshot: string
+    @NonEmptyProperty()
+    state: string
+    @OptionalProperty()
+    link: string | undefined
+    @OptionalNumberArrayOf()
+    scores: number[] | undefined
+    @OptionalProperty()
+    votes: number 
+}
+
+export class SnapshotProposalLoader extends GraphQLDataLoaderBase<Proposals, Proposal> {
+    public info = INFO;
+
+    protected cursorMode = "cursor" as const;
+    protected cursor: number;
+
+    protected batchSize = BATCH_SIZE;
+    protected type = Proposal;
+    protected cadenceConfig = {
+        [ScheduleMode.BACKFILL]: { seconds: 5 },
+        [ScheduleMode.INCREMENTAL]: { minutes: 5 },
+    };
+
+    protected graphQLQuery: DocumentNode;
+    protected codec = ProposalsCodec;
+
+    constructor(client: GraphQLClient,space: String) {
+        super(client);
+        this.graphQLQuery = makeQUERY(space)
+    }
+
+    protected preLoad(
+        context: InternalLoadContext
+    ): TE.TaskEither<ProgramError, InternalLoadContext> {
+        // Ensure a number cursor is used
+        if (!context.cursor) {
+            context.cursor = 0
+        }
+        if (typeof context.cursor == "string") {
+            context.cursor = Number(context.cursor)
+        }
+
+        this.cursor = context.cursor
+        
+        return TE.right(context);
+    }
+
+    protected mapResult(result: Proposals): Array<Proposal> {
+        return result.proposals as Array<Proposal>
+    }
+
+    protected extractCursor(proposals: Proposals) {
+        return String(this.cursor+proposals.proposals.length);
```

Comment:
I wonder how scalable their solution of `first`/`skip` will be. Unfortunately this kind of pagination has a complexity of `O(n)` so we can expect this to slow down later down the road. At that point we might want to refactor this somehow.

Reply:
The other option I found here for cursor is with the created_gt field. My initial implementation used this, so I could easily convert it to that, but I figured this was more robust as there could be multiple proposals with the same created value.

Reply:
Nah, it is fine, but we should keep an eye on this as it might become a problem in the future.